### ISSUE TITLE
set req.length 

### ISSUE THREAD

**Issue by dcolens** [expressjs/body-parser#214]
it would be nice that body-parser sets the length of the body, my use case is I'd like to know the length of a JSON payload POSTED using chunked-encoding, so there is no content-length header and req.body is set to an Object thanks to body-parser.

Sounds like the only way to get the body size is to stringify req.body and get the length :/ or use raw-parser and JSON.parse.

Any chance we could set req.length or similar to the size of the raw body before parsing it ? (other suggestions are welcome)

**Comment by dougwilson** [comment]
That sounds mostly OK, but we need to impact what setting a `length` property to `req` will have against all the existing application using this module out there. That property seems likely to be already in-use, but maybe not. Help determining one way or the other would be very awesome :) !

**Comment by dcolens** [comment]
Agreed about the name conflict, don't see a way to validate the it upfront.

What about `req.bodyLength`, even if it is used, it is likely for the same reason :)

**Comment by jordonias** [comment]
Can you just use the `verify` option?

```
bodyParser.json({
  verify: function(req, res, buf, encoding) {
    req.bodyLength = buf.length;
  }
})
```

**Comment by dougwilson** [comment]
Hi @dcolens I'm circling back around on this, and was wondering if you could say what your use-case is. The reason is not to dismiss this request, but rather to determine the best way to implement the request. Basically looking to better understand the "why?" to it to make sure everything lines up once implemented.

**Comment by dcolens** [comment]
I receive a JSON body, in some cases I want to forward it to another service and I need to stringify it, but I only want to do that if its size is less than 100KB. The only way to test it, is to JSON.stringify it which is blocking (I know that JSON.parse is too), I'd rather check the size before calling JSON.stringify.

**Comment by dougwilson** [comment]
@dcolens gotcha. What do you do if the size is over 100KB?

**Comment by dcolens** [comment]
POST it to a remote service, i.e. JSON.stringify

**Comment by dougwilson** [comment]
Oh, I thought it said you didn't want to send it to the remove service if it was over 100KB?

**Comment by dcolens** [comment]
Sorry, misunderstood, indeed, in that case I omit the incoming payload from the outgoing call to the remote service! 
The idea is that some calls are tracked, but if the user sends too much payload, he does not benefit from the tracking.

**Comment by dougwilson** [comment]
Makes sense! And to clarify, in all the cases you do actually look at the body locally, just the size is what determines if you want to forward it?

**Comment by dcolens** [comment]
yes pretty much

**Comment by rmullenix** [comment]
Any update on this?

**Comment by jonchurch** [comment]
We can generalize this issue down to a request for:

**Accessing the raw request body in further middleware**

This is a bit of a seperation of concerns issue, in that body-parser's job is to present a parsed `req.body` downstream of it in the middleware chain. I say seperation of concerns because body-parser inherently abstracts away dealing with raw bodies outside of its link in the middleware chain. 

So the natural suggestion would be to encapsulate that raw body access into its own middleware, e.g. :
```js
function rawBodyBuffer((req, res, next) => {
  req.rawBody = ''
  // this does not play nice with body-parser
  req.on('data', (chunk) => req.rawBody += chunk)
  req.on('end', () => { next() })
})
```
but there is long standing tension w/ body-parser in that approach because body-parser expects to consume the body stream. 

Placing `rawBodyBodyBuffer` before body-parser in the middleware chain consumes body data before body-parser can touch it (breaking body-parser's ability to parse the body at all), and placing it after body-parser means that the body will have been consumed before `rawBodyBuffer` runs (breaking rawBodyBuffer's ability to consume the body).


### `verify`'s role

You could very well consider `verify` to actually be `inspectRaw` because that is the hook body-parser provides for inspecting (verifying) the raw body before committing to parsing. 

It's your last shot to deal with that buffer before it is gobbled by the parser (assuming you're not using `bodyParser.raw()` which is a a pass through bufferer). This isn't made particularly clear in the [documentation](https://github.com/expressjs/body-parser#verify).

it's not _great_ to use it add a new property to the `req`, but it is definitely the best bet you have to both parse the body and capture its raw length without recomputing it from the parsed body.
